### DISCUSSION TITLE
SML: Bug? Able to read one variable from a SML energy meter but not a second!

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#16981] [Q&A] [answered]
Hi all, I seem to have found a strange bug, not sure what to make out of it:

Im reading a SML/OBIS energy meter with Tasmota 11.1.0, meter definitions are:

    >M 1
    +1,3,s,0,9600,WUA
    1,77070100600100ff@#,Zählernummer,,zaehlerid,0
    1,77070100020800ff@1000,Lieferung,kWh,liefer,0
    1,77070100010800ff@1000,Bezug,kWh,bezug,0
    #

The values **do** get parsed and displayed, so that's working. I get

    SENSOR = {"Time":"2022-11-02T15:22:27","WUA":{"zaehlerid":"0a014c475a00032fd4f2","liefer":22192,"bezug":59242}}

Now I want to parse the two SML variables into local variables before sending them via WebSend, like so

    >D
    bez=0
    lief=0
    >B
    =>sensor53 r
    tper=60
    >S
    if upsecs%tper==0
    then
    lief=WUA#liefer
    bez=WUA#bezug
    print lief: %lief% bez: %bez%
    => WebSend [receiver.example.com] /?id=wua-main&time=%time%&uptime=%uptime%&bezug=%bez%&lieferung=%lief%
    endif

but I only get this

    15:28:40.456 RSL: SENSOR = {"Time":"2022-11-02T15:28:40","WUA":{"zaehlerid":"0a014c475a00032fd4f2","liefer":22192,"bezug":59242}}
    15:29:31.462 var not found: WUA#liefer
    15:29:31.464 lief: 0.00 bez: ???
    15:29:31.466 Script: performs "WebSend [receiver.example.com] /?id=wua-main&time=929.00&uptime=5.00&bezug=???&lieferung=0.00"
    15:29:31.558 RSL: RESULT = {"WebSend":"Done"}

That proves that the SML interface is working (see first log line) but then Tasmota complains about not finding WUA#liefer while it does NOT complain about WUA#bezug . It sets lief to 0.00 but bez to ???

Half an hour ago I could read "lief" for a short period of time but not now, never "bez".  Not enough space somewhere?

Any ideas?

Andreas

**Comment by sfromis** [comment]
Well, you have no trouble *reading* both values via SML. What you then can parse from the JSON is where the issue is. Related thread: #16943

**Reply by hagenbuch** [reply]
Thanks a lot, but I don't use MQTT on that device. Can I parse JSON without using MQTT, too?

**Reply by barbudor** [reply]
All exemples I see with similar syntax such as BME280#Temperature are in the >T section which makes sense as this is the section that receive the SENSOR message
I think you should separate the reading and the sending
```
>T
lief=WUA#liefer
bez=WUA#bezug

>S
if upsecs%tper==0
then
print lief: %lief% bez: %bez%
=> WebSend [receiver.example.com] /?id=wua-main&time=%time%&uptime=%uptime%&bezug=%bez%&lieferung=%lief%
endif
```

**Reply by sfromis** [reply]
And yes, the JSON parsing is not related to MQTT being connected (or not) for the publish.

**Comment by hagenbuch** [comment, marked as answer]
Thank you all for your replies, but I managed to get everything together: it took me a while to find out that the extra line in the >D section is needed!

For English language speakers: "Bezug" means consumption, "Lieferung" means production or delivery.

    >D
    lief=0
    bez=0
    zz=""
    # this is important!
    >B
    =>sensor53 r
    tper=60
    >T
    bez=WUAPV#bezug
    lief=WUAPV#lieferung
    zz=WUAPV#zaehlerid
    >S
    if upsecs%tper==0
    then
    print zz: %zz% lief: %lief% bez: %bez% atime: %tstamp% up: %uptime%
    => WebSend [example.com] /?id=mydeviceid&atime=%tstamp%&uptime=%uptime%&bezug=%bez%&lieferung=%lief%
    endif
    >M 1
    +1,3,o,0,300,WUAPV,1,100,2F3F210D0A
    1,0.0.0(@#),Zählernummer,,zaehlerid,0
    1,1.8.0(@1,Bezug,kWh,bezug,1
    1,2.8.0(@1,Lieferung,kWh,lieferung,1
    1,31.7.0(@1,Strom L1,A,l1,2
    1,51.7.0(@1,Strom L2,A,l2,2
    1,71.7.0(@1,Strom L3,A,l3,2
    #

I am reading an Itron EM214 counter here.
